### PR TITLE
BFD with Static Route - Automation

### PR DESCRIPTION
**Pull request by harsgoll**
### Description of PR
This PR adds automation for below test cases

1. test_bfd_with_lc_reboot_ipv4
2. test_bfd_static_route_deletion_ipv4
3. test_bfd_with_lc_reboot_ipv6
4. test_bfd_static_route_deletion_ipv6
5. test_bfd_flap_ipv4
6. test_bfd_flap_ipv6
7. test_bfd_with_rp_reboot_ipv4
8. test_bfd_remote_link_flap_ipv4
9. test_bfd_with_rp_reboot_ipv6
10. test_bfd_remote_link_flap_ipv6
11. test_bfd_lc_asic_shutdown_ipv4
12. test_bfd_lc_asic_shutdown_ipv6
13. test_bfd_portchannel_member_flap_ipv4
14. test_bfd_config_reload_ipv4
15. test_bfd_portchannel_member_flap_ipv6
16. test_bfd_with_rp_config_reload_ipv4
17. test_bfd_with_rp_config_reload_ipv6
18. test_bfd_with_bad_fc_asic_ipv4
19. test_bfd_config_reload_ipv6
20. test_bfd_with_bad_fc_asic_ipv6

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

#### How did you verify/test it?
Verified on T2 Q100, Q200 testbed (internal to cisco)

#### Any platform specific information?
Supports Q100, Q200

### CODE REVIEW

**Comment by siqbal1986** [comment]
please add reviewers

File: tests/common/plugins/conditional_mark/tests_mark_conditions.yaml [hunk 1]
```diff
@@ -64,6 +64,12 @@ arp/test_wr_arp.py:
 #######################################
 #####            bfd              #####
 #######################################
+bfd/test_bfd_static_route.py:
+  skip:
+    reason: "Only supported on chassis system."
+    conditions:
+      - "is_multi_asic is False"
```
Comment (harsgoll) [inline tests/common/plugins/conditional_mark/tests_mark_conditions.yaml:71]: Change the condition to cisco and T2.

File: tests/bfd/test_bfd_static_route.py [hunk 2]
```diff
@@ -0,0 +1,1369 @@
...
+from bfd_base import BfdBase
+import logging
+import time
+from tests.common.utilities import wait_until
+from tests.common.config_reload import config_reload
+from tests.common.platform.processes_utils import wait_critical_processes
+from tests.common.reboot import reboot
+
+pytestmark = [
+    pytest.mark.topology('t2')
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:11]: Remove the marker as it is already covered in skip conditions.

File: tests/bfd/test_bfd_static_route.py [hunk 3]
```diff
@@ -0,0 +1,1369 @@
...
+                    c. If expected state is "Up" and no. of down peers is 0, output is True
+                    d. If expected state is "Down" and no. of up peers is 0, output is True
+        """
+        try:
+            duts = duthosts.frontend_nodes
+            for dut in duts:
+                bfd_base_instance.modify_all_bfd_sessions(dut, "false")
+            for dut in duts:
+                # config reload
+                #config_reload(dut)
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:39]: Uncomment this

File: tests/bfd/test_bfd_static_route.py [hunk 4]
```diff
@@ -0,0 +1,1369 @@
...
+        # Savings the configs
+        src_dut.shell('sudo config save -y')
+        
+        # Perform a cold reboot on source dut
+        reboot(src_dut, localhost)
+
+        # Waiting for all processes on Source dut
+        wait_critical_processes(src_dut)
+
+        check_bgp = request.getfixturevalue("check_bgp")
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:89]: Check if iBGP is also verified here

File: tests/bfd/test_bfd_static_route.py [hunk 5]
```diff
@@ -0,0 +1,1369 @@
...
+        results = check_bgp()
+        failed = [result for result in results if "failed" in result and result["failed"]]
+        if failed:
+            pytest.fail("BGP check failed, not all BGP sessions are up. Failed: {}".format(failed))
+
+        check_bgp = request.getfixturevalue("check_bgp")
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:95]: Looks duplicate. Check and remove it.

File: tests/bfd/test_bfd_static_route.py [hunk 6]
```diff
@@ -0,0 +1,1369 @@
...
+            if self.test_case_status:
+                successful_iterations += 1
+        
+        # Determine the success rate
+        logger.info("successful_iterations: %d", successful_iterations)
+        success_rate = (successful_iterations / self.total_iterations) * 100
+        
+        logger.info("Current success rate: %.2f%%", success_rate)
+        # Check if the success rate is above the threshold (e.g., 98%)
+        assert success_rate >= 98, "BFD flap verification success rate is below 98% ({}%)".format(success_rate)
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:421]: Make it 100 to uncover any flakiness if present.

File: tests/bfd/test_bfd_static_route.py [hunk 7]
```diff
@@ -0,0 +1,1369 @@
...
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:488]: Make it 100%

File: tests/bfd/test_bfd_static_route.py [hunk 8]
```diff
@@ -0,0 +1,1369 @@
...
+        request.config.portchannels_on_dut = "src"
+        request.config.selected_portchannels = list_of_portchannels_on_src
+        
+        #Shutdown PortChannel members
+        for portchannel_interface in list_of_portchannels_on_src:
+            action = "shutdown"
+            list_of_portchannel_members_on_src = bfd_base_instance.extract_backend_portchannels(src_dut)[portchannel_interface]['members']
+            request.config.selected_portchannel_members = list_of_portchannel_members_on_src
+            for each_member in list_of_portchannel_members_on_src:
+                self.control_interface_state(src_dut, src_asic, bfd_base_instance, each_member, action)
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:877]: Potential enhancement:
When 1 member goes down, BFD should go down (2 member port channel - Q200) as well as for 3 member port channel. Verify BFD behaviour.

File: tests/bfd/test_bfd_static_route.py [hunk 9]
```diff
@@ -0,0 +1,1369 @@
...
+        # Waiting for all processes on Source & destination dut
+        wait_critical_processes(src_dut)
+        wait_critical_processes(dst_dut)
+
+        check_bgp = request.getfixturevalue("check_bgp")
+        results = check_bgp()
+        failed = [result for result in results if "failed" in result and result["failed"]]
+        if failed:
+            pytest.fail("BGP check failed, not all BGP sessions are up. Failed: {}".format(failed))
+
```
Comment (harsgoll) [inline tests/bfd/test_bfd_static_route.py:1058]: Transition verification required.
Add log analyzer regex to ensure DOWN string is observed in syslogs.

**Comment by abdosi** [comment]
@harsgoll : please fix Pre_test Static Analysis

**Comment by abdosi** [comment]
@harsgoll static analysis still failing. Can you refer above steps and run locally and then commit the changes ?

**Comment by abdosi** [comment]
@harsgoll : still issue with Static Analysis

**Comment by harsgoll** [comment]
@abdosi : I have addressed all static analysis related errors.
Cc: @vperumal

**Review by abdosi — APPROVED** [review]

**Pull request merged**